fqdn: Pre-allocate ToFQDN selector identities

gandro opened the pull request:
This PR contains a further performance improvement for https://github.com/cilium/cilium/pull/32769, namely pre-allocation for selector identities as proposed in [CFP-28427](https://github.com/cilium/design-cfps/blob/fe6dd4faf9c03626428e63536d0f5c5ec6e82ef1/cilium/CFP-28427-fqdn-selectors.md#fqdn-identity-pre-allocation).

The basic idea is that we pre-allocate a matching identity whenever a new ToFQDN selector is added by the policy engine. This reduces critical path latency for DNS lookups, as no policy maps need to be updated when the first DNS lookup occurs.

See commit messages for details

squeed commented on pkg/fqdn/identities.go:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	lbls := labels.FromSlice([]labels.Label{selector.IdentityLabel()})
+	id, _, err := a.backend.AllocateLocalIdentity(lbls, true, identity.IdentityUnknown)
```
How bad would it be to pass `notifyOwners: false`, then batch all updates to the SelectorCache ourselves?

gandro replied:
I'll look into it

squeed replied:
On second thought, let's not do this. We might wind up in a bad race condition where IPs get an identity that is not yet plumbed in.

We should keep `notifyOwner: true`

gandro committed:
identity: Add AllocateLocalIdentity to interface

This adds the recently added AllocateLocalIdentity function to the
IdentityAllocator interface, allowing interface users to call it as
well. In addition, implement the function on the mock implementation.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
fqdn: Pre-allocate FQDN selector identities

This commit introduces logic in the FQDN name manager to pre-allocate
identities when new selectors are added. If the pre-allocated selector
does not overlap with other selectors, then this reduces critical path
latency for DNS lookups, as no policy maps need to be updated when
the first DNS lookup occurs.

The downside of this approach is that the BPF policy map entry might
never be used if DNS names are matched by multiple selectors or if a
selector never matches any DNS names.

Actual pre-allocation and release of identities are done out-of-band, as
allocating an identity calls back into the `SelectorCache` and would
case deadlocks if we did it inline. Most of the code added in this
commit therefore is the infrastructure to enqueue and process allocation
and release requests in a Go routine running in the background.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
labels: Export `world` labels

They will be used out of package in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
